### ISSUE TITLE
[BUG] Service Bus Processor update disposition timeout error

### ISSUE THREAD

**Issue by liukun-msft** [Azure/azure-sdk-for-java#30483]
**Describe the bug**
User encountered `Update disposition request timed out` error while setting `maxConcurrentCalls = 50` and `prefetchCount = 80`.

***Exception or Stack Trace***
Add the exception log and stack trace if available
```
14:01:04.441 [reactor-executor-1] DEBUG c.a.c.a.i.handler.ReceiveLinkHandler - {"az.sdk.message":"onDelivery.","connectionId":"MF_82b18a_1660543257822","errorCondition":null,"errorDescription":null,"entityPath":"testqueue","linkName":"testqueue_d5aa87_1660543257878","updatedLinkCredit":63,"remoteCredit":0,"delivery.isPartial":false,"delivery.isSettled":false}
14:02:01.107 [parallel-14] DEBUG c.a.m.s.i.ServiceBusReactorReceiver - linkName[testqueue_d5aa87_1660543257878]: Cleaning timed out update work tasks.
14:03:01.112 [parallel-14] DEBUG c.a.m.s.i.ServiceBusReactorReceiver - linkName[testqueue_d5aa87_1660543257878]: Cleaning timed out update work tasks.
14:03:01.112 [parallel-14] ERROR c.a.m.s.i.ServiceBusReactorReceiver - Update disposition request timed out., errorContext[NAMESPACE: servicebusliuku.servicebus.windows.net. ERROR CONTEXT: N/A, PATH: testqueue, REFERENCE_ID: testqueue_d5aa87_1660543257878, LINK_CREDIT: 63]
com.azure.core.amqp.exception.AmqpException: Update disposition request timed out., errorContext[NAMESPACE: servicebusliuku.servicebus.windows.net. ERROR CONTEXT: N/A, PATH: testqueue, REFERENCE_ID: testqueue_d5aa87_1660543257878, LINK_CREDIT: 63]
	at com.azure.messaging.servicebus.implementation.ServiceBusReactorReceiver.lambda$cleanupWorkItems$17(ServiceBusReactorReceiver.java:368)
	at java.base/java.util.concurrent.ConcurrentHashMap.forEach(ConcurrentHashMap.java:1603)
	at com.azure.messaging.servicebus.implementation.ServiceBusReactorReceiver.cleanupWorkItems(ServiceBusReactorReceiver.java:359)
	at com.azure.messaging.servicebus.implementation.ServiceBusReactorReceiver.lambda$new$0(ServiceBusReactorReceiver.java:95)
	at reactor.core.publisher.LambdaSubscriber.onNext(LambdaSubscriber.java:160)
	at reactor.core.publisher.FluxInterval$IntervalRunnable.run(FluxInterval.java:125)
	at reactor.core.scheduler.PeriodicWorkerTask.call(PeriodicWorkerTask.java:59)
	at reactor.core.scheduler.PeriodicWorkerTask.run(PeriodicWorkerTask.java:73)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**To Reproduce**

Processor Client
```Java
ServiceBusProcessorClient processorClient = new ServiceBusClientBuilder()
        .connectionString(Credentials.serviceBusConnectionString)
        .processor()
        .queueName(Credentials.serviceBusQueue)
        .processMessage(processMessage)
        .processError(processError)
        .prefetchCount(80)
        .maxConcurrentCalls(50)
        .maxAutoLockRenewDuration(Duration.ofSeconds(0))
        .disableAutoComplete()
        .buildProcessorClient();
```

Process message 
```Java
Consumer<ServiceBusReceivedMessageContext> processMessage = messageContext -> {
    try {
        System.out.println(messageContext.getMessage().getMessageId());
        
        //Sleep 1 second to mock process progress
        TimeUnit.SECONDS.sleep(1);
        
        messageContext.complete();
    } catch (Exception ex) {
        messageContext.abandon();
    }
};
```

Important step: **add vm option: -Dreactor.schedulers.defaultBoundedElasticSize=50**

Users add this vm option to increase the boundedElasticSize on the AKS pod. if they do not explicitly add this option, the default BoundedElasticSize will be `10 * number of processors`, i.e. 10.

```Java
public static final int DEFAULT_BOUNDED_ELASTIC_SIZE = (Integer)Optional.ofNullable(
    System.getProperty("reactor.schedulers.defaultBoundedElasticSize")).map(Integer::parseInt)
.orElseGet(() -> {
    return 10 * Runtime.getRuntime().availableProcessors();
});
```
**Expected behavior**

**Setup (please complete the following information):**
- service bus version 7.10.0
- AKS pod with **one processor**

**Comment by liukun-msft** [comment]
**Reason**
The ServiceBusProcessorClient hangs because we have other tasks running on the `Schedulers.boundedElastic` thread pool in addition to the receive message tasks. **When we limit the size of the thread pool to the concurrency number**, it gets stuck when all threads are used in processing messages, this is because it can't create new thread for other tasks. Therefore, we need to set defaultBoundedElasticSize to a value greater than the concurrency. (I change it to at least 54 and solve the problem.)

Below are some places we use Schedulers.boundedElastic pool:

- processor client health monitor
    ```Java
    //ServiceBusProcessorClient#start
    monitorDisposable = Schedulers.boundedElastic().schedulePeriodically(() -> {
        if (this.asyncClient.get().isConnectionClosed()) {
            restartMessageReceiver(null);
        }
    }, SCHEDULER_INTERVAL_IN_SECONDS, SCHEDULER_INTERVAL_IN_SECONDS, TimeUnit.SECONDS);
    ```

- process message (maxConcurrentCalls > 1)
    ```Java
    //ServiceBusProcessorClient#receiveMessages
    receiveFlux.parallel(processorOptions.getMaxConcurrentCalls(), 1)
        .runOn(Schedulers.boundedElastic(), 1)
        .subscribe(subscribers);
    ````
- receive message (receive from ReactorReceiver, reactor-executor thread)
    ```Java
    //ServiceBusReactorReceiver
    public Flux<Message> receive() {
        return super.receive()
            .filter(message -> message != EMPTY_MESSAGE)
            .publishOn(Schedulers.boundedElastic());
    }
    ```
- receive message (receive from ServiceBusReactorReceiver, Schedulers.boundedElastic() thread)
    ```Java
    //ServiceBusReceiveLinkProcessor
    next.receive().publishOn(Schedulers.boundedElastic()).subscribe(...)
    ```

- create receive link
    ```Java
    //ServiceBusReceiveLinkProcessor
    next.getEndpointStates().subscribeOn(Schedulers.boundedElastic()).subscribe(...)
    ```

There are some other places in AMQP package which is to close `ReactorReceiver` and `RequestResponseChannel`.

**Current Solution**
If `maxConcurrentCalls` > 10 * number of processors, set vm option: -Dreactor.schedulers.defaultBoundedElasticSize greater than `maxConcurrentCalls`.

**Comment by liukun-msft** [comment]
**Pending Question**
- Why are no threads completed and release thread resources to the pool?
   This is probably because a thread running on "Schedulers.boundedElastic" may create another task that also requires a thread from "Schedulers.boundedElastic" and there are no threads in the pool. This is like a deadlock, holding a thread and trying to get another thread from the pool. As a result, no threads will be completed and the thread resources released, and then the client is hung for a timeout.

- Why no `ReactorRejectedExecutionException` be thrown if thread pool is full according to [code](https://github.com/reactor/reactor-core/blob/main/reactor-core/src/main/java/reactor/core/scheduler/BoundedElasticScheduler.java#L733)?
Pending investigate.

- Some users may not configure for defaultBoundedElasticSize, so the default value is `10 * number of processors`, which will be a limit to their maxConcurrentCalls. Maybe we need to create our own boundedElastic pool based on maxConcurrentCalls and maxConcurrentSessions.
   TODO item.

**Comment by liukun-msft** [comment]
Root cause:
1. credit calculation issue.
2. The backpressure from reactor-operators-chain cause the "disposition ack" stay in receive-link handler buffer and wait timeout.

**Comment by YogeshwarPrajapatiTR** [comment]
@liukun-msft, i have one listener can say one processor , meaning that i can use Max 10 bounded BoundedElastic deafult.

Question 1) If i set Max maxConcurrentCalls 20 as per below code even thought it will use 10 bounded thread?
Question 2) And if i want to use more Bounded Thread then have to add vm option : Dreactor.schedulers.defaultBoundedElasticSize=20. this understanding is correct?
Question 3) what is the right combination for using prefetch and Maxconcurrent call ,where can get high performance

`ServiceBusProcessorClient processorClient = new ServiceBusClientBuilder()
        .connectionString(Credentials.serviceBusConnectionString)
        .processor()
        .queueName(Credentials.serviceBusQueue)
        .processMessage(processMessage)
        .processError(processError)
        .prefetchCount(10)
        .maxConcurrentCalls(20)
        .disableAutoComplete()
        .buildProcessorClient();`

**Comment by anuchandy** [comment]
This is addressed, currently shipped as beta

```xml
<dependency>
    <groupId>com.azure</groupId>
    <artifactId>azure-messaging-servicebus</artifactId>
    <version>7.15.0-beta.4</version>
</dependency>
```

**Comment by wstest123** [comment]
@anuchandy The attached version is beta version, could you please help advice when will the GA verson ready for us to use in PROD environment? Thanks!

**Comment by robsonkades** [comment]
We are going through the same problems here.
@anuchandy, is there any estimated release date?

**Comment by anuchandy** [comment]
Hello @wstest123, @robsonkades, this is planned to be GA-ed in next 3 weeks.

**Comment by anuchandy** [comment]
Shipped 7.15.0. Note that the original issue was due to not draining the disposition acknowledgement from the service, this is addressed. On the other hand, if application queues up disposition requests in a rate higher than the host can dequeue and send over the wire, causing the queue to grow, then the timeout can still happen.

**Comment by anuchandy** [comment]
Resolving this. Refer https://learn.microsoft.com/en-us/azure/developer/java/sdk/troubleshooting-messaging-service-bus-overview#upgrade-to-715x for details about version 7.15. Also refer previous [comment](https://github.com/Azure/azure-sdk-for-java/issues/30483#issuecomment-1930775396).
